CoreScriptsRoot/Modules/NewChat.lua

```diff
@@ -0,0 +1,233 @@
+--[[
+	// FileName: NewChat.lua
+	// Written by: Xsitsu
+	// Description: Bridges the topbar in corescripts to any chat system running in the non-corescripts environment.
+]]
+local CoreGuiService = game:GetService("CoreGui")
+local RobloxGui = CoreGuiService:WaitForChild("RobloxGui")
+
+local StarterGui = game:GetService("StarterGui")
+local GuiService = game:GetService("GuiService")
+
+local Util = require(RobloxGui.Modules.ChatUtil)
+
+
+local moduleApiTable = {}
+do
+		local ChatWindowState = 
+		{
+			Visible = true,
+			MessageCount = 0,
+			TopbarEnabled = true,
+			CoreGuiEnabled = true,
+		}
+
+		local communicationsConnections = {}
+		local eventConnections = {}
+
+		local MakeSystemMessageCache = {}
+
+		local function FindIndexInCollectionWithType(collection, indexName, type)
```

Comment:
This name is misleading, because you are finding the Value stored and not the index; `FindInCollectionByKeyAndType` would read a lot better